### PR TITLE
chore: merge 3.6

### PR DESCRIPTION
**Pull request by ycliuhw**
Merge branch '3.6' into main:
- https://github.com/juju/juju/pull/17994
- https://github.com/juju/juju/pull/17999
- https://github.com/juju/juju/pull/18009
- https://github.com/juju/juju/pull/17980
- https://github.com/juju/juju/pull/18019
- https://github.com/juju/juju/pull/18057
- https://github.com/juju/juju/pull/17960
- https://github.com/juju/juju/pull/18070
- https://github.com/juju/juju/pull/18081
- https://github.com/juju/juju/pull/18083
- https://github.com/juju/juju/pull/18091
- https://github.com/juju/juju/pull/18080
- https://github.com/juju/juju/pull/18085
- https://github.com/juju/juju/pull/18103
- https://github.com/juju/juju/pull/18100
- https://github.com/juju/juju/pull/18073
- https://github.com/juju/juju/pull/18123
- https://github.com/juju/juju/pull/18111
- https://github.com/juju/juju/pull/18105
- https://github.com/juju/juju/pull/18124
- https://github.com/juju/juju/pull/18128
- https://github.com/juju/juju/pull/18052
- https://github.com/juju/juju/pull/18095
- https://github.com/juju/juju/pull/18126
- https://github.com/juju/juju/pull/17699
- https://github.com/juju/juju/pull/18022
- https://github.com/juju/juju/pull/18154
- https://github.com/juju/juju/pull/18155
- https://github.com/juju/juju/pull/18163
- https://github.com/juju/juju/pull/18118
- https://github.com/juju/juju/pull/18152

```
# Conflicts:
#       .github/workflows/static-analysis.yml
#       agent/addons/addons.go
#       agent/addons/addons_test.go
#       agent/introspect/introspect_test.go
#       api/apiclient_test.go
#       api/client/charms/client_test.go
#       apiserver/facades/client/application/deploy_test.go
#       apiserver/facades/client/applicationoffers/applicationoffers.go
#       apiserver/facades/client/applicationoffers/applicationoffers_test.go
#       cmd/containeragent/initialize/package_test.go
#       cmd/juju/action/common.go
#       cmd/juju/action/run.go
#       cmd/juju/action/showoperation.go
#       cmd/juju/action/showtask.go
#       cmd/juju/controller/config.go
#       cmd/juju/metricsdebug/collectmetrics.go
#       cmd/jujud/agent/agenttest/agent.go
#       cmd/jujud/agent/caasoperator.go
#       cmd/jujud/agent/machine.go
#       cmd/jujud/main.go
#       controller/config.go
#       core/cache/controller.go
#       database/txn/transaction.go
#       featuretests/dblog_test.go
#       go.mod
#       go.sum
#       internal/mongo/mongo_test.go
#       internal/mongo/service.go
#       internal/upgrades/upgradevalidation/upgrade.go
#       internal/worker/diskmanager/diskmanager.go
#       internal/worker/introspection/worker.go
#       internal/worker/uniter/paths.go
#       internal/worker/uniter/paths_test.go
#       internal/worker/uniter/relation/resolver.go
#       internal/worker/uniter/relation/resolver_test.go
#       internal/worker/uniter/runner/context/context_test.go
#       internal/worker/uniter/runner/context/contextfactory.go
#       mongo/mongotest/mongoservice_mock.go
#       state/backups/files.go
#       state/backups/files_test.go
#       state/stateenvirons/features.go
#       tests/includes/colors.sh
#       tests/suites/secrets_iaas/juju.sh
#       tests/suites/secrets_iaas/vault.sh
#       tests/suites/secrets_k8s/k8s.sh
#       tests/suites/static_analysis/lint_go.sh
#       worker/modelcache/worker_test.go
#       worker/uniter/runner/context/contextfactory_test.go
#       worker/uniter/runner/jujuc/add-metric.go
#       worker/uniter/runner/jujuc/add-metric_test.go
#       worker/uniter/runner/jujuc/k8s-raw-get_test.go
#       worker/uniter/runner/jujuc/k8s-raw-set_test.go
#       worker/uniter/runner/jujuc/k8s-spec-get_test.go
#       worker/uniter/runner/jujuc/k8s-spec-set.go
#       worker/uniter/runner/jujuc/k8s-spec-set_test.go
```

### CODE REVIEW

**Comment by ycliuhw** [comment]
Hey @SimonRichardson @jack-w-shaw @manadart @Aflynn50 
this forward merge has conflicts that were introduced on your end, so it would be best for you to review it.

**Review by SimonRichardson — APPROVED** [review]
This seems ok.

**Review by Aflynn50 — APPROVED** [review]
The changes from me in this look good

File: doc/MongoDB-and-Consistency.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
+<!-- TODO(gfouillet): do not merge into 4.0, or delete whenever merged (reason: related to mongodb) -->
```
Comment (manadart) [inline doc/MongoDB-and-Consistency.md:1]: Address this.

**Review by manadart — COMMENTED** [review]
My stuff (relation hook resolver) looks OK.

File: doc/MongoDB-and-Consistency.md (hunk 1 above)
Comment (ycliuhw) [reply doc/MongoDB-and-Consistency.md]: Yep, removed.

**Review by hpidcock — CHANGES_REQUESTED** [review]
what is with the git history here being re-written?

**Comment by hpidcock** [comment]
I will force merge this after review, static-analysis job is too pedantic.

**Review by hpidcock — APPROVED** [review]

**Pull request merged**